File: src/_03_BehavioralPattern/_03_01_ChainOfResponsibilities/RequestHandler.js
/**
 * @typedef { import("./Request").default } Request
 */

export default class RequestHandler {
  /** @type { RequestHandler } */
  #nextHandler;

  /** @param { RequestHandler } nextHandler */
  constructor(nextHandler) {
    this.#nextHandler = nextHandler;
  }

  /** @param { Request } request */
  handler(request) {
    if(this.#nextHandler) this.#nextHandler.handler(request);
  }
}

File: src/_03_BehavioralPattern/_03_01_ChainOfResponsibilities/after/RequestHandler.java
package _03_BehavioralPattern._03_01_ChainOfResponsibilities.after;

public abstract class RequestHandler {

  private RequestHandler nextHandler;

  public RequestHandler(RequestHandler nextHandler) {
    this.nextHandler = nextHandler;
  }

  public void handler(Request request) {
    if(this.nextHandler != null) this.nextHandler.handler(request);
  }
  
}
